constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/IntervalMap.java

```diff
@@ -0,0 +1,437 @@
+          rightValue = Optional.of(rightNextDefinedSegment.value());
+          nextRightIndex = rightIndex + 1;
+        } else {
+          rightInterval = Interval.between(
+              Duration.MIN_VALUE,
+              Inclusive,
+              rightNextDefinedSegment.interval().start,
+              rightNextDefinedSegment.interval().startInclusivity.opposite());
+          rightValue = Optional.empty();
+        }
+      } else {
+        rightInterval = Interval.FOREVER;
+        rightValue = Optional.empty();
+      }
+
+      if (leftInterval.end.isEqualTo(rightInterval.end)) {
+        endTime = leftInterval.end;
+        if (leftInterval.includesEnd() && rightInterval.includesEnd()) {
+          endInclusivity = Inclusive;
+          leftIndex = nextLeftIndex;
+          rightIndex = nextRightIndex;
+        } else if (leftInterval.includesEnd()) {
+          endInclusivity = Exclusive;
+          rightIndex = nextRightIndex;
+        } else if (rightInterval.includesEnd()) {
+          endInclusivity = Exclusive;
+          leftIndex = nextLeftIndex;
+        } else {
+          endInclusivity = Exclusive;
+          rightIndex = nextRightIndex;
+        }
+      } else if (leftInterval.end.shorterThan(rightInterval.end)) {
+        endTime = leftInterval.end;
+        endInclusivity = leftInterval.endInclusivity;
+        leftIndex = nextLeftIndex;
+      } else {
+        endTime = rightInterval.end;
+        endInclusivity = rightInterval.endInclusivity;
+        rightIndex = nextRightIndex;
+      }
+      var finalInterval = Interval.between(startTime, startInclusivity, endTime, endInclusivity);
+      if (finalInterval.isEmpty()) continue;
+
+      var newValue = transform.apply(leftValue, rightValue);
+      if (newValue.isPresent()) {
+        if (!newValue.equals(previousValue)) {
+          result.add(Segment.of(finalInterval, newValue.get()));
+        } else {
+          var previousInterval = result.remove(result.size() - 1).interval();
+          result.add(
+              Segment.of(
+                  Interval.unify(previousInterval, finalInterval),
+                  newValue.get()
+              )
+          );
+        }
+      }
+      previousValue = newValue;
+      startTime = endTime;
+      startInclusivity = endInclusivity.opposite();
+    }
+
+    return resultMap;
+  }
+
+  /**
+   * The left bound of the leftmost defined interval.
+   * @return the leftmost bound and its inclusivity
+   */
+  public Optional<Pair<Duration, Interval.Inclusivity>> minValidTimePoint() {
+    if(!this.isEmpty()) {
+      final var window = this.segments.get(0).interval();
+      return Optional.of(Pair.of(window.start, window.startInclusivity));
+    } else{
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * The right bound of the rightmost defined interval.
+   * @return the rightmost bound and its inclusivity
+   */
+  public Optional<Pair<Duration, Interval.Inclusivity>> maxValidTimePoint() {
```

Comment:
I left it in because I thought there might be some later, if we ever use IntervalMap for more general profiles. But I can't think of any, so I don't mind removing it.